13006 rewrite reason for appointment page tests

LanceGundersen opened the pull request:
## Description
We want to rewrite our Enzyme based tests with tests in RTL that do not test implementation details.

## Testing done
Local and unit

## Screenshots
![image](https://user-images.githubusercontent.com/8315447/94297643-d5e19580-ff32-11ea-918a-d58a1672285f.png)


## Acceptance criteria
- [ ] Test coverage matches
- [ ] RTL Tests fully replace enzyme tests

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

LanceGundersen committed:
Add setTypeOfFacility mock

LanceGundersen committed:
Rewrite reasonForAppointment tests

jbalboni commented on src/applications/vaos/tests/new-appointment/components/ReasonForAppointmentPage.unit.spec.jsx:
```diff
@@ -1,189 +1,236 @@
-        updateReasonForAppointmentData={updateReasonForAppointmentData}
-        data={{}}
-      />,
+    fireEvent.click(
```
The logic for choosing the next page after this one isn't conditional on the options, so I don't think you need to test that each option takes you to the right next page.

LanceGundersen replied:
Understood

jbalboni commented on src/applications/vaos/tests/new-appointment/components/ReasonForAppointmentPage.unit.spec.jsx:
```diff
@@ -1,189 +1,236 @@
-    form.unmount();
-  });
+    const textBox = screen.getByRole('textbox');
+    fireEvent.change(textBox, { target: { value: 'test' } });
```
Since this text box is optional for CC, it would be good to have a test that continued to the next page without entering anything. I think you could repurpose this test for that.

LanceGundersen replied:
Makes sense

jbalboni commented on src/applications/vaos/tests/new-appointment/components/ReasonForAppointmentPage.unit.spec.jsx:
```diff
@@ -1,189 +1,236 @@
-    const history = {
-      push: sinon.spy(),
-    };
+  it('should show validation for VA medical request', async () => {
```
One other thing to note with these validation messages is that there's some hidden functionality as far as what the character limit for the reason field is. If you get to this page as part of the direct schedule flow, then the character limit changes depending on the radio option that you choose (because we concatenate that value and the reason info into one field to send to the backend).

I think the easiest way to set the flow would be to dispatch this action: https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/vaos/new-appointment/redux/actions.js#L197

So like `store.dispatch(startDirectSchedulFlow())` in your test before the render.

LanceGundersen replied:
Gotcha, implemented the dispatch in the beforeEach call.

jbalboni reviewed (commented):
Looks great so far. There's some other functionality on this page we should test, I made a comment that should help you trigger that behavior.

LanceGundersen committed:
Fix tests based on feedback

- Removed unnecessary click events
- Removed duplicate text box testing
- Added dispatch to beforeEach

jbalboni commented on src/applications/vaos/tests/new-appointment/components/TypeOfFacilityPage.unit.spec.jsx:
```diff
@@ -24,11 +25,16 @@ const initialState = {
+  beforeEach(() => {
+    mockFetch();
+    store = createTestStore(initialState);
+    store.dispatch(startDirectScheduleFlow());
```
What I meant in that comment was that you can write a test in the reason for appointment page test that dispatches this action, renders the reason for appointment page, and verifies that the character limit for the reason box changes depending on which radio option you choose. You shouldn't need to modify this test.

LanceGundersen replied:
Resolved, the naming of the test could probably be better but I believe this is the test you are seeking.

LanceGundersen committed:
Add test for character limit change

- reverts accidental facilty page test
